File: sms/msg91.go
package sms

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

type MSG91Provider struct {
	AuthKey    string
	TemplateID string
	SenderID   string
	HTTPClient *http.Client
}

func NewMSG91Provider(authKey, templateID, senderID string) *MSG91Provider {
	return &MSG91Provider{
		AuthKey:    authKey,
		TemplateID: templateID,
		SenderID:   senderID,
		HTTPClient: &http.Client{Timeout: 10 * time.Second},
	}
}

func (p *MSG91Provider) SendOTP(ctx context.Context, payload SMSPayload) error {
	if p.AuthKey == "" {
		return fmt.Errorf("msg91 auth key missing")
	}

	url := fmt.Sprintf("https://control.msg91.com/api/v5/otp?template_id=%s&mobile=%s&authkey=%s", p.TemplateID, payload.Mobile, p.AuthKey)
	
	reqBody, err := json.Marshal(map[string]string{
		"OTP": payload.OTP,
	})
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBuffer(reqBody))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("msg91 API error: status code %d", resp.StatusCode)
	}

	return nil
}

func (p *MSG91Provider) GetProviderName() string {
	return "MSG91"
}

File: backend_api/sms/msg91.py
import requests
from sms.provider import SMSProvider, SMSPayload

class MSG91Provider(SMSProvider):
    def __init__(self, auth_key: str, template_id: str, sender_id: str):
        self.auth_key = auth_key
        self.template_id = template_id
        self.sender_id = sender_id

    def send_otp(self, context, payload: SMSPayload):
        if not self.auth_key:
            raise ValueError("msg91 auth key missing")

        url = f"https://control.msg91.com/api/v5/otp?template_id={self.template_id}&mobile={payload.mobile}&authkey={self.auth_key}"
        
        data = {"OTP": payload.otp}

        response = requests.post(
            url,
            json=data,
            headers={"Content-Type": "application/json"},
            timeout=10
        )

        if response.status_code >= 400:
            raise ValueError(f"msg91 API error: status code {response.status_code}")

    def get_provider_name(self) -> str:
        return "MSG91"
